Source/JavaScriptCore/builtins/JSIteratorConstructor.js

```diff
@@ -71,3 +71,47 @@ function from(value)
     // 6. Return wrapper.
     return @wrapForValidIteratorCreate(iterator, iteratorNextMethod);
 }
+
+// https://tc39.es/proposal-iterator-sequencing/#sec-iterator.concat
+function concat(/* ...items */)
+{
+    "use strict";
+
+    var count = @argumentCount();
+
+    var openMethods = [];
+    var iterables = [];
+    for (var i = 0; i < count; ++i) {
+        var iterable = arguments[i];
+        if (!@isObject(iterable))
+            @throwTypeError("Iterator.concat expects all arguments to be objects");
+
+        var openMethod = iterable.@@iterator;
+        if (@isUndefinedOrNull(openMethod))
+            @throwTypeError("Iterator.concat expects all arguments to be iterable");
+
+        @arrayPush(openMethods, openMethod);
+        @arrayPush(iterables, iterable);
+    }
+
+    var generator = (function*() {
+        for (var i = 0; i < count; ++i) {
+            var iterator = openMethods[i].@call(iterables[i]);
+            if (!@isObject(iterator))
+                @throwTypeError("Iterator.concat expects all arguments to be iterable");
+
+            for (;;) {
+                var result = @iteratorGenericNext(iterator.next, iterator);
```

Comment:
Please hoist `iterator.next` out of the loop, it is observable and per spec performed only once per iterator.